### PR TITLE
[Gutenberg] Add amp-fit-text support to text blocks

### PR DESCRIPTION
**Pull request by miina**
Adds `amp-fit-text` toggle and controls to the following blocks:
- paragraph;
- heading;
- subheading;
- code;
- quote;

Settings when amp-fit-text is selected:
screen shot 2018-05-16 at 9 16 08 pm

Currently has the default layout of `fixed-height`, perhaps this could be changed with a next iteration once AMP Layout has been implemented.

### CODE REVIEW

**Comment by westonruter** [comment]
Conflicts need resolving.

**Comment by miina** [comment]
Develop merged & conflicts resolved.

File: assets/js/amp-editor-blocks.js [hunk 1]
```diff
@@ -314,6 +345,77 @@ var ampEditorBlocks = ( function() {
...
+			minFont = props.attributes.minFont,
+			maxFont = props.attributes.maxFont,
+			height = props.attributes.height,
+			isSelected = props.isSelected,
+			el = wp.element.createElement,
+			InspectorControls = wp.blocks.InspectorControls,
+			TextControl = wp.components.TextControl,
+			ToggleControl = wp.components.ToggleControl,
+			PanelBody = wp.components.PanelBody,
+			label = 'Use AMP Fit Text';
```
Comment (westonruter) [inline assets/js/amp-editor-blocks.js]: Needs translation

File: assets/js/amp-editor-blocks.js [hunk 2]
```diff
@@ -314,6 +345,77 @@ var ampEditorBlocks = ( function() {
...
+				el( InspectorControls, { key: 'inspector' },
+					el( PanelBody, { title: component.data.ampSettingsLabel },
+						el( ToggleControl, {
+							label: label,
+							checked: ampFitText,
+							onChange: function() {
+								props.setAttributes( { ampFitText: ! ampFitText } );
+							}
+						} ),
+						el( TextControl, {
```
Comment (westonruter) [inline assets/js/amp-editor-blocks.js]: Can the `font-size-picker` control be used instead?

File: assets/js/amp-editor-blocks.js [hunk 3]
```diff
@@ -314,6 +345,77 @@ var ampEditorBlocks = ( function() {
...
+							value: minFont,
+							onChange: function( nextMinFont ) {
+								props.setAttributes( { minFont: nextMinFont } );
+							}
+						} ),
+						el( TextControl, {
+							label: __( 'Max font (px)' ),
+							value: maxFont,
+							onChange: function( nextMaxFont ) {
+								props.setAttributes( { maxFont: nextMaxFont } );
```
Comment (westonruter) [inline assets/js/amp-editor-blocks.js]: Make sure the `height` is set to be the `Math.max()` of the `nextMaxFont` and its current value.

File: assets/js/amp-editor-blocks.js [hunk 4]
```diff
@@ -314,6 +345,77 @@ var ampEditorBlocks = ( function() {
...
+							value: height,
+							onChange: function( nextHeight ) {
+								props.setAttributes( { height: nextHeight } );
+							}
+						} ),
+						el( TextControl, {
+							label: __( 'Min font (px)' ),
+							value: minFont,
+							onChange: function( nextMinFont ) {
+								props.setAttributes( { minFont: nextMinFont } );
```
Comment (westonruter) [inline assets/js/amp-editor-blocks.js]: Abort if `nextMinFont` is larger than `maxFont`

File: assets/js/amp-editor-blocks.js [hunk 5]
```diff
@@ -314,6 +345,77 @@ var ampEditorBlocks = ( function() {
...
+
+		if ( ampFitText ) {
+			return isSelected && (
+				el( InspectorControls, { key: 'inspector' },
+					el( PanelBody, { title: component.data.ampSettingsLabel },
+						el( ToggleControl, {
+							label: label,
+							checked: ampFitText,
+							onChange: function() {
+								props.setAttributes( { ampFitText: ! ampFitText } );
```
Comment (westonruter) [inline assets/js/amp-editor-blocks.js]: Should `ampFitText` be replaced with `event.checked`?

Comment (westonruter) [inline assets/js/amp-editor-blocks.js]: Can this also result in the `font-size-picker` and `range-control` under Text Settings to be hidden when `ampFitText` is true?

Comment (miina) [reply assets/js/amp-editor-blocks.js]: Most of the core blocks seem to use negating the current value for toggle controls (as far as I've seen), thought of staying consistent with the core blocks when using that. Thoughts?

**Comment by miina** [comment]
@westonruter Changes made. It's still possible to set larger minFont than maxFont, for example when first setting `minFont` to 30 and then setting `maxFont` to something lower. However, adding the check for aborting `maxFont` change when it's lower than `minFont` would make editing the values very unfriendly for the user. E.g. if `maxFont` and `minFont` are both 15 and the user would like to set `maxFont` to 25, then removing `5` from `15` wouldn't be possible since then the temporary `1` would be lower than the `minFont` value.

Would be great if the InspectorControls would have validation option (for example with option to display error message) and disable saving in case of incorrect values instead of automatically changing other values. Wondering if something like this is planned, tried to find an issue for it but without success.

**Comment by westonruter** [comment]
@pbakaus with 937035f I've made it so that if you are using an AMP block in content that the required AMP component scripts will get added to the page even in non-AMP documents. This ensures that the components actually render instead of being empty, which seems better. The downside here is that the AMP runtime is trying to take control of the entire page, not just the components you're using. So for example attempting to submit the comment form results in an error:

image

I think that the AMP runtime should perhaps skip doing anything other than render specific components when the document lacks the boilerplate or the `amp` attribute on the `html` element. So that the AMP components can be better used as a component library. Would this be “dirty AMP”?

**Comment by pbakaus** [comment]
@westonruter good call. Yes, I think it's fair to file this on the amphtml repo as issue, should be a fairly easy fix.

**Comment by westonruter** [comment]
> I think it's fair to file this on the amphtml repo as issue

https://github.com/ampproject/amphtml/issues/15583

**Pull request merged**